move Windows logs to system

fearful-symmetry opened the pull request:
## What does this PR do?

This PR is meant to provide a functional OOTB logging experience for Integrations on windows. More precisely, we're moving most of the logs from the Windows integration (save for the powershell logs) over to the system integration. Right now, there's no OOTB logging on windows, as users would have to manually add the Windows integration. I'm not a windows expert, and I'm still working on testing this with actual data, but right now the system integration installs everything and isn't broken pre-ingest.

This is also sort of a draft, as I want other people's input as to what I've migrated and if there's certain log data streams that we either want to migrate back to windows, or add to system.

## Checklist

- [X] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/master/CONTRIBUTING.md#tips-for-building-integrations) and this pull request is aligned with them.
- [x] I have verified that all datasets collect metrics or logs.
- [x] I've updated the docs

## How to test this PR locally

- I tested this with https://github.com/elastic/elastic-package
- Once you spin everything up, and have an agent running on windows, we're looking to make sure:
  - All three windows log sources (application, system, security) are collected on windows
  -  The agent isn't trying to start the log sources on non-windows hosts. Look for a "module X not found" error in the agent logs.

## Related issues

- https://github.com/elastic/integrations/issues/399

## Screenshot
<img width="1040" alt="Screen Shot 2020-12-10 at 11 43 26 AM" src="https://user-images.githubusercontent.com/8418476/101821569-fbffe600-3adc-11eb-988f-f9c3d2252dfe.png">

<img width="1294" alt="Screen Shot 2020-12-10 at 11 53 07 AM" src="https://user-images.githubusercontent.com/8418476/101822529-4b92e180-3ade-11eb-99a3-45f097f99cef.png">

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-411/detail/PR-411/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-411/detail/PR-411/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-411/detail/PR-411/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-411/detail/PR-411/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://integrations_411.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #411 opened]`

* Start Time: 2020-11-20T20:49:54.182+0000

* Duration: 9 min 9 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 10  |
| Skipped      | 0 |
| Total        | 10   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->





<!--COMMENT_GENERATED_WITH_ID_comment.id-->

leehinman commented on dev/import-beats-resources/windows/docs/README.md:
```diff
@@ -49,17 +42,3 @@ The Windows `powershell_operational` dataset provides events from the Windows
-
-{{fields "security"}}
-
-### Sysmon/Operational
```
Do we want to move Sysmon?  You have to install Sysmon on the system first before you get these events.  Also am I right in assuming another PR will add the "Application" and "System" event logs?

jamiehynds replied:
Correct Lee, Sysmon should not be included in the System package. The end result should look like this. 

- **System**: Application, System, Security
- **Windows:** Sysmon, Powershell, Forwarded Events + any new Windows integrations we build in the future.
- **Windows Custom:** Allow users to ingest events from any other channel.

Given that Windows events will be included across multiple packages, we need ensure users have some guidance around this. @leehinman suggested adding a 'Windows' category in Fleet, which makes a lot of sense.

fearful-symmetry replied:
Ah, I didn't know that. My goal here was just to have something equivalent to the logging streams we have on everything else. Do you have any other suggestions? Should we keep the sysmon stuff in windows?

fearful-symmetry replied:
@narph do you have any input here?

fearful-symmetry replied:
@jamiehynds didn't see your comment.

> System: Application, System, Security

Do we have `application` and `system` data streams in the windows integration now? Is this something in winlogbeat that we need to add to Integrations?

narph replied:
my preference would be to have all logs bundled in one package, if we want to have only the out of the box/default logs in the `system` package then indeed going for `application`, `security` and `system` in the system package and keeping the rest inside the `windows` package seems the way to go. Regarding `application` and `system` data streams, I have added some clarifications below, I think the data streams were even added at some point the windows package but removed after, maybe we can find that PR if it makes things easier.

narph replied:
https://github.com/elastic/integrations/pull/281

leehinman replied:
> @jamiehynds didn't see your comment.
> 
> > System: Application, System, Security
> 
> Do we have `application` and `system` data streams in the windows integration now? Is this something in winlogbeat that we need to add to Integrations?

`Application` and `System` are not currently in the windows integration, but they were.  This PR has what it looked like elastic/integrations#281 They are very simple, no javascript processing.

leehinman commented:
I think the visualizations will need to be updated as well, they have things like 

```
"query":"data_stream.dataset:windows.security"
```
and now that the data_stream is different this will have to be updated.  I think we want to do an OR so the visualization will still work even if the data was ingested with the old package vs the new.

elasticmachine commented:
Pinging @elastic/security-external-integrations (Team:Security-External Integrations)

elasticmachine commented:
Pinging @elastic/integrations-services (Team:Services)

mtojek reviewed (changes requested):
As there were incompatible changes pushed to the master branch (PR: https://github.com/elastic/integrations/pull/410), please rebase this PR against the master branch.

Long story short: `dev/import-beats-resources/windows/docs/README.md` is `packages/windows/_dev/build/docs/README.md` now.

fearful-symmetry commented:
@leehinman Yep, completely forgot about the searches. Gonna do that next.

ph commented:
@fearful-symmetry the condition part should work, I've created https://github.com/elastic/package-spec/issues/86 to add it to the spec. I think we were more relax in validation before and we haven't added the condition key after.

fearful-symmetry commented:
So @jamiehynds I think we need a better inventory of what exactly we need/and want in here for 7.12, since I'm not a huge windows expert here. Winlogbeat only has `sysmon`, `security` and `powershell` modules. You mentioned application and system logs. Where are those coming from?

jamiehynds commented:
> So @jamiehynds I think we need a better inventory of what exactly we need/and want in here for 7.12, since I'm not a huge windows expert here. Winlogbeat only has `sysmon`, `security` and `powershell` modules. You mentioned application and system logs. Where are those coming from?

@fearful-symmetry we don't have modules for Application and System, but users can still ingest the events despite not being fully mapped to ECS. In winlogbeat a user can just add the 'application' or 'system' event channels in the winlogbeat.yml. Example here: https://github.com/elastic/beats/blob/7de72d68c5ceaca4ae326bc0eba072b3baecaaf8/x-pack/winlogbeat/winlogbeat.yml#L21-L57

narph commented:
> So @jamiehynds I think we need a better inventory of what exactly we need/and want in here for 7.12, since I'm not a huge windows expert here. Winlogbeat only has `sysmon`, `security` and `powershell` modules. You mentioned application and system logs. Where are those coming from?

just to add  a clarification here @fearful-symmetry  the application and system logs are and have been a part of  Winlogbeat logs, they were removed from the windows package last moment for 7.10 and the idea was to have them introduced in the system package instead, @jamiehynds correct me if I am wrong here.

jamiehynds commented:
Correct @narph! We (security integrations) weren't aware that application and system events were planned for the system package so pulled them just before the 7.10 release. Subsequently, we all agreed that the app/system/security events should reside in the system package.

fearful-symmetry commented:
Thanks for all the input @narph @jamiehynds ! So, just to make sure I have the right idea:

Migrate from the windows integration:
- security

Add to the system integration: 
- Application logs
- system log

fearful-symmetry commented:
Alright, this is still a work in progress and the dashboards are still broken, but otherwise everything should at least be in the right place.

jamiehynds commented:
> Alright, this is still a work in progress and the dashboards are still broken, but otherwise everything should at least be in the right place.

Thanks @fearful-symmetry! We'll need to consider how to transition security events from the Windows package for users who enabled it in 7.10, but @leehinman and I will discuss.

fearful-symmetry commented:
Okay, so one thing we need to figure out is if we're renaming the data streams. `application`,  `system` and `security` are pretty generic, and if we have these windows-only data sets in with everything else, it might be confusing. I'm also worried about compatibility/upgrade issues if we rename the data streams to `windows_application` and so on.

mukeshelastic commented:
Let's keep the datastreams as they were from windows package. The intent is to integrate data collection and visualization the datastreams is an implementation detail that doesn't need to be same especially if they contain different set of fields between OSes. I know renaming the datastreams later will be hard but I don't see a reason for doing that at this point other than more work on operators side to manage data lifecycle as now they have to enable ILM policies on more datastreams than a single. 

@fearful-symmetry feel free to work with @leehinman on validating that the windows dashboards load correctly in 7.11 after the moving the data ingestion from windows package to system.

fearful-symmetry commented:
> Let's keep the datastreams as they were from windows package. The intent is to integrate data collection and visualization the datastreams is an implementation detail that doesn't need to be same especially if they contain different set of fields between OSes. I know renaming the datastreams later will be hard but I don't see a reason for doing that at this point other than more work on operators side to manage data lifecycle as now they have to enable ILM policies on more datastreams than a single.

Yah, I think you're right. Even if the names feel awkward we should probably keep them the same. However, we'll still have to deal with the fact we're dealing with `system.*` and not `windows.*`

fearful-symmetry commented:
okay, made a few changes. @mukeshelastic are we sure we want to keep the data stream names the same? My concern is that something like `system.system` could be confusing for users.

mukeshelastic commented:
@fearful-symmetry Just to make sure I followed you right, the windows system events will be called system.system events after they are migrated off windows package onto system package? If yes, then that does look confusing when you need to act on it.. like applying ILM policies at more granular level ( i.e. system.system rather than system.*) I am hoping we can solve this problem by tooltip UX in ILM editor or documentation.  

@ruflin @andresrc @urso @ph thoughts?

urso commented:
> thoughts?

The system package will collect metrics and logs. In order to make the difference more clear for users managing the data streams we might want to reflect the data-type (metrics/logs) in the name to some extend? E.g. 

`system.system -> system.logs`
`system.application -> system.application_logs`
`system.security -> system.security_logs`

The datastream names for metrics we keep as is.

P.S.: I wonder if we could support a naming schema like `system.logs.application`, `system.logs.security`, `system.metrics.cpu`, ...

ruflin commented:
@urso The dataset part can be whatever you want as long as it doesn't contain a `-`.

ph commented:
` system.application -> system.application_logs`

@urso You don't need the `logs` part, this is the type and should already prefix the data stream like `logs-nginx.access-production`

urso commented:
> You don't need the logs part, this is the type and should already prefix the data stream like logs-nginx.access-production

🤦

fearful-symmetry commented:
So, right now I'm going with `windows_application` and so on.

Right now I'm erring on the side of just renaming them to remote the prefixes and just do `system.application` and so on, and use other UI tips to clarify things for users as @mukeshelastic suggested.

fearful-symmetry commented:
So, from an Integrations UI perspective, removing any `windows_` prefixes from the data stream names works. I think this will mostly be an issue of how users feel about it from a data perspective. 

<img width="1048" alt="Screen Shot 2020-12-01 at 10 53 33 AM" src="https://user-images.githubusercontent.com/8418476/100784084-7cc72f80-33c3-11eb-904d-c83bd26b1057.png">

fearful-symmetry commented:
Also, my dynamic input doesn't seem to be working:
```
2020-12-01T18:48:46Z: type: 'STATE': sub_type: 'RUNNING' message: Application: filebeat--7.11.0-SNAPSHOT[c13ae1e0-b3c0-4de6-8b5f-f27c522f1e48]: State changed to DEGRADED: 1 error: 3 errors: Error creating runner from config: Error creating input. No such input type exist: 'winlog'; Error creating runner from config: failed in processor.javascript: GoError: Could not load module 'windows': cannot load windows, only built-in modules are supported; Error creating runner from config: Error creating input. No such input type exist: 'winlog'
```

ycombinator commented:
/retest

ycombinator commented:
jenkins, run the tests please

fearful-symmetry commented:
Well, CI is passing

mukeshelastic commented:
> Also, my dynamic input doesn't seem to be working:
> 
> ```
> 2020-12-01T18:48:46Z: type: 'STATE': sub_type: 'RUNNING' message: Application: filebeat--7.11.0-SNAPSHOT[c13ae1e0-b3c0-4de6-8b5f-f27c522f1e48]: State changed to DEGRADED: 1 error: 3 errors: Error creating runner from config: Error creating input. No such input type exist: 'winlog'; Error creating runner from config: failed in processor.javascript: GoError: Could not load module 'windows': cannot load windows, only built-in modules are supported; Error creating runner from config: Error creating input. No such input type exist: 'winlog'
> ```

were you able to resolve this @fearful-symmetry

fearful-symmetry commented:
@mukeshelastic I think I'm still running into it:
```
2020-12-02T19:10:34Z: type: 'STATE': sub_type: 'RUNNING' message: Application: filebeat--7.11.0-SNAPSHOT[8b422dc4-3a35-4a44-b2d0-b05af37936ae]: State changed to DEGRADED: 1 error: 3 errors: Error creating runner from config: Error creating input. No such input type exist: 'winlog'; Error creating runner from config: failed in processor.javascript: GoError: Could not load module 'windows': cannot load windows, only built-in modules are supported; Error creating runner from config: Error creating input. No such input type exist: 'winlog'
```

Rebased against master, updated the `elastic-package` tool.

fearful-symmetry commented:
@urso What's this mean?

` cannot load windows, only built-in modules are supported;`

Trying to figure out if this is something else or if the conditional still isn't working.

fearful-symmetry commented:
@ycombinator do we know what the dynamic inputs are supposed to "look like" once they're rendered?

I'm looking at the final config in the Kibana UI, and I don't see any conditionals:

```
      - id: winlog-system.application-7d959394-0f42-45ee-b547-0ba435fe62e8
        name: Application
        data_stream:
          dataset: system.application
          type: logs
        ignore_older: 72h
      - id: winlog-system.security-7d959394-0f42-45ee-b547-0ba435fe62e8
        name: Security
        data_stream:
          dataset: system.security
          type: logs
```

urso commented:
> Trying to figure out if this is something else or if the conditional still isn't working.

Looks like you get this error from the javascript processor: https://github.com/elastic/beats/blob/master/libbeat/processors/script/javascript/module/require/require.go#L36

Looks like the filebeat binary is missing this package: https://github.com/elastic/beats/blob/master/libbeat/processors/script/javascript/module/windows/windows.go

The includes of the javascript processor are here (windows is included): https://github.com/elastic/beats/blob/master/libbeat/processors/script/javascript/module/include.go

The "windows" Javascript module is only compiled into windows builds. Either you try to configure the winlog input (or use some JS) on Linux/Mac by accident, or for some reason the filebeat build is missing the javascript imports.

fearful-symmetry commented:
> The "windows" Javascript module is only compiled into windows builds. Either you try to configure the winlog input (or use some JS) on Linux/Mac by accident, or for some reason the filebeat build is missing the javascript imports.

Okay, so probably still an issue with the input guard then. Thanks!

ph commented:
@fearful-symmetry The _condition_ should be there @nchaulet or @jfsiii does the config builder also need an update because of https://github.com/elastic/package-spec/pull/87 ? if this is the case Lets get a fix in to unblock @fearful-symmetry .

nchaulet commented:
Yes we will have, to update the config build to add the condition to the config, I can make the change but just want to be sure I understand it correctly.

We want the `condition` part of the package input to be in the agent policy as it is right?

ph commented on packages/system/manifest.yml:
```diff
@@ -33,6 +33,10 @@ policy_templates:
       - type: logfile
         title: Collect logs from System instances
         description: Collecting System auth and syslog logs
```
You probably want to exclude to run on windows.
```
${host.platform} != 'windows'
```

ph commented:
@nchaulet Yes correct, can you create an issue and take this.

@fearful-symmetry Can you add the complete generate YAML from your configuration I just want to make sure we generate the right structure based on the manifest.

fearful-symmetry commented:
@ph the entire yml is over 2200 lines, so I'm just putting it here if you want it. There's a shorter snippet above, but I'm not entirely sure where the conditional ends up, I just assume it's in the data_stream settings.
[elastic-agent(2).yml.txt](https://github.com/elastic/integrations/files/5637348/elastic-agent.2.yml.txt)

fearful-symmetry commented:
Brief update. So, I've tested this on Windows and things seem file. Aside from the conditionals, this appears to be fine. Do we want to merge this now, since we're so close to the FF, or wait for the blockers to wrap up and test again? @mukeshelastic

mukeshelastic commented:
@fearful-symmetry I am unclear on the risk of "conditionals" poses to getting the system integration work as expected on windows to collect windows logs. @urso @ph @ruflin @andresrc thoughts?

ruflin commented:
I have not read the full thread and all the references, but I wonder why this is a blocker. If the condition would be on the stream level, I would assume it just works?

ph commented:
As mentioned in https://github.com/elastic/package-spec/issues/86#issuecomment-740352800 the conditional shouldn't be a blocker, @nchaulet and @ycombinator suggestion allow you to use them today.

fearful-symmetry commented:
Okay, I moved the conditionals to the stream files. I'm still trying to test this, as `elastic-package` seems to be having issues. I'd like to at least keep things moving through.

fearful-symmetry commented:
It's working! 

```
      - id: winlog-system.application-705828c0-d18f-4ed3-82f5-3868645e369d
        name: Application
        data_stream:
          dataset: system.application
          type: logs
        condition: '${host.platform} == ''windows'''
        ignore_older: 72h
```

fearful-symmetry commented:
Okay, it seems like we're ready to merge. Anything I'm missing?

ph commented:
I've looked at the _conditon_ part and they look fine for me. But I havent' tested the integration locally on my machine.

fearful-symmetry commented:
Alright. Re-tested the actual conditionals themselves...seems fine to me?

leehinman commented on packages/windows/manifest.yml:
```diff
@@ -1,6 +1,6 @@
 name: windows
 title: Windows
-version: 0.2.12
+version: 0.3.0
```
I think we need to set the `kibana.version` below the same as was done for the system `kibana.version`

leehinman reviewed (changes requested):
Looks great, just think we need the `kibana.version` set for the windows package, so we force the correct pairing of the versions of system & windows integrations.

leehinman reviewed (approved):
LGTM
